code/game/machinery/telecomms/machine_interactions.dm

```diff
@@ -27,74 +29,202 @@
-		dat += "<br>Network: <a href='?src=[REF(src)];input=network'>[network]</a>"
-		dat += "<br>Prefabrication: [autolinkers.len ? "TRUE" : "FALSE"]"
-		if(hide)
-			dat += "<br>Shadow Link: ACTIVE</a>"
-
-		//Show additional options for certain machines.
-		dat += Options_Menu()
-
-		dat += "<br>Linked Network Entities: <ol>"
-
-		var/i = 0
-		for(var/obj/machinery/telecomms/T in links)
-			i++
-			if(T.hide && !hide)
-				continue
-			dat += "<li>[REF(T)] [T.name] ([T.id])  <a href='?src=[REF(src)];unlink=[i]'>\[X\]</a></li>"
-		dat += "</ol>"
-
-		dat += "<br>Filtering Frequencies: "
-
-		i = 0
-		if(length(freq_listening))
-			for(var/x in freq_listening)
-				i++
-				if(i < length(freq_listening))
-					dat += "[format_frequency(x)] GHz<a href='?src=[REF(src)];delete=[x]'>\[X\]</a>; "
-				else
-					dat += "[format_frequency(x)] GHz<a href='?src=[REF(src)];delete=[x]'>\[X\]</a>"
-		else
-			dat += "NONE"
+/obj/machinery/telecomms/ui_interact(mob/user, datum/tgui/ui)
+	operator = user
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Telecomms")
+		ui.open()
+
+/obj/machinery/telecomms/ui_data(mob/user)
+	var/list/data = list()
+
+	data += add_option()
+
+	data["minfreq"] = MIN_FREE_FREQ
+	data["maxfreq"] = MAX_FREE_FREQ
+	data["frequency"] = tempfreq
+
+	var/obj/item/multitool/heldmultitool = get_multitool(user)
+	data["multitool"] = heldmultitool
+
+	if(heldmultitool)
+		data["multibuff"] = heldmultitool.buffer
+
+	data["toggled"] = toggled
+	data["id"] = id
+	data["network"] = network
+	data["prefab"] = autolinkers.len ? TRUE : FALSE
+
+	var/list/linked = list()
+	var/i = 0
+	data["linked"] = list()
+	for(var/obj/machinery/telecomms/machine in links)
+		if(machine.hide && !hide)
+			continue
+		i++			//Original was above the check if machine is hidden, index revealed there are more machines
+		var/list/entry = list()
+		entry["index"] = i
+		entry["name"] = machine.name
+		entry["id"] = machine.id
+		linked += list(entry)
+	data["linked"] = linked
+
+	var/list/frequencies = list()
+	data["frequencies"] = list()
+	for(var/x in freq_listening)
+		frequencies += list(x)
+	data["frequencies"] = frequencies
+
+	var/list/circuits = list()
+	var/n = 0
+	data["circuits"] = list()
+	for(var/c in GLOB.ic_speakers)
+		var/obj/item/integrated_circuit/I = c
+		var/obj/item/O = I.get_object()
+		var/list/circuit = list()
+		if(get_area(O)) //if it isn't in nullspace, can happen due to printer newing all possible circuits to fetch list data
+			n++
+			circuit["index"] = n
+			circuit["name"] = O.name
+			circuit["coords"] = "[O.x], [O.y], [O.z]"
+			circuits += list(circuit)
+	data["circuits"] = circuits
+
+	return data
+
+/obj/machinery/telecomms/ui_act(action, params)
+	if(..())
+		return
 
-		dat += "<br>  <a href='?src=[REF(src)];input=freq'>\[Add Filter\]</a>"
-		dat += "<hr>"
+	if(!issilicon(usr))
+		if(!istype(usr.get_active_held_item(), /obj/item/multitool))
+			return
```

Comment:
```suggestion
	if(!issilicon(operator) && !istype(operator.get_active_held_item(), /obj/item/multitool))
		return
```

Using `usr` (it really is typed as `/mob`) here is mostly okay, but let's be safe and consistent.

(Speaking of consistency, you probably *can* replace `operator` with `usr` if you *really* want.)